web/handlers.go

```diff
@@ -343,6 +352,34 @@ func (h Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	}
 }
 
+func (h *Handler) checkScopes(c *Context, appID string) model.Scope {
+	oAuthApp, err := c.App.GetOAuthApp(appID)
```

Comment:
@larkox - is it always guaranteed that `oAuthApp` is non-nil (or, in general, any obj. that comes from the DB), or is it always the case that `oAuthApp` nil means that `err` is non-nil? We've seen issues like : https://github.com/mattermost/mattermost-server/pull/18437, trying to prevent similar issue here.

Reply:
In theory, if there is a entry in the database, it will return the entry (with all the information), and if not, it will return a "not found" error. That being said, that is also the case for what is being solved in the other PR.

It will probably not hurt to add the check. @agnivade any thought on this?

Reply:
Yes, it is guaranteed that if err is nil, then oAuthApp is not nil. It is a well-establlished convention and I don't think there is a need to again check for `oAuthApp == nil`.

#18437 was _not_ that case. It happened because we were ignoring not found errors. I have removed all those redundant checks now - https://github.com/mattermost/mattermost-server/pull/18440.

Reply:
Thanks @agnivade , that makes sense, it was my understanding of the "theory" as well = ). What is less clear (to me) is the "guaranteed" aspect of the "well-established convention" and how do we make sure that the convention is well-enforced throughout the codebase and for new PRs. In the  #18437 case, fwis, the code was structured in such a way that the convention was not respected (the `appErr` could be nil  and `referencedPost` could've been nil as well). In this case, having the extra checks, while reducing the readability, etc. , would've prob. avoided the havoc it created. Bottom line, it looks like the q is: either we rely on the developers to be aware of AND enforce the convention (or find other - automatic? - ways to ensure it is observed) or we'd need to add defensive code with the assumption that the convention _could_ not be respected (with the corresponding drawbacks).  Not sure where the answer lies, but the potential for future mishap has not been eliminated yet imo.

Reply:
Thanks Catalin, I totally get your point. For sure, there are automated [checks](https://github.com/golang/tools/tree/master/go/analysis/passes/nilness) we already run to catch these cases, but they are not always 100% effective.

> Bottom line, it looks like the q is: either we rely on the developers to be aware of AND enforce the convention (or find other - automatic? - ways to ensure it is observed)

IMO, adding defensive coding is also a human factor. How can you "guarantee" that every developer always does defensive coding? The answer should be IMO, to write tests to lock the behavior. But as we all know well enough, that doesn't always happen due to "reasons". I'd suggest for PR reviewers to continue the good practice of asking authors to add unit tests to guarantee these API contracts.

We already have tests which guarantee this behavior for all store methods. But some cases like these, as we can see, the error handling is not done properly. We need to be more careful in PR reviews and add more tests.

Tests and automated checks are the best possible tools against these problems. We just need to spend the time to write them.

Reply:
Thanks @agnivade , that makes sense, it was my understanding of the "theory" as well = ). Going beyond the scope of this particular change, what is less clear (to me) is the "guaranteed" aspect of the "well-established convention" and how do we make sure that the convention is well-enforced (aka guaranteed) throughout the codebase and for new PRs. You are right, #18437 does not illustrate the case (though one could argue that it illustrates a failure to fully understand the other "side" of the "convention": when appErr is not-nil, the referencedPost is nil) , but the point imo still stands -  we still rely on the developers to be aware of AND enforce the convention (e.g. in cases other than "get obj from store"). With the assumption  that the convention _could_ not be observed (or interpreted incorrectly as above) , one would need to add defensive code - with the corresponding drawbacks - or fix the code - which might be too late.  Not sure where the answer lies, but the potential for future mishap is still there imo.